README.md
# PyConFR 2019 - GRPC NLP

## Instructions

Structure is based on [this article](https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure). Source code can be found in the `src` folder, and tests in the `tests` folder.

### Installation

To install the package (development mode):

```bash
➤ make pip-install
```

### Tests

#### Tox
~~We use `tox` for the tests. This ensure a clear separation between the development environment and the test environment.
To launch the tests, run the `tox` command:~~

~~It first starts with a bunch of checks (`flask8` and others) and then launch the tests using python 3.~~

#### Pytest
You can use `pytest` for the tests:
```bash
➤ make pytest           
pytest -v
```


history.md
## next release
 
## 0.1.0
(2019-10-19)

- tools: packaging tools using for the project.

tests/test_run_twint_with_multiprocessing.py
import logging
import math
import multiprocessing
from collections import defaultdict

import pytest
import queue
from dataclasses import dataclass
from typing import Iterator, List

import twint
from pyconfr_2019.grpc_nlp.protos.StorageService_pb2 import StoreTweetsResponse, StoreTweetsRequest

from storage.processors.twint_with_multiprocess import run_twint_with_multiprocessing

cache_multiprocessing = {}


@pytest.fixture
def build_twint_configs():
    def _build_twint_configs(twitter_users=("PyConFr", "PyCon")):
        twint_configs = []
        twint_limit = 25
        twint_debug = False
        for twitter_user in twitter_users:
            twint_config = twint.Config()
            twint_config.Username = twitter_user
            twint_config.Limit = twint_limit  # bug with the Limit parameter not working only factor of 25 tweets
            twint_config.Debug = twint_debug
            twint_configs.append(twint_config)
        return twint_configs

    return _build_twint_configs


def test_run_twint_with_multiprocessing(build_twint_configs, mocker, caplog):
    nb_tweets_by_chunk = 20
    twitter_users = ("PyConFr", "PyCon", "ThePSF")
    twint_configs = build_twint_configs(twitter_users)
    twint_limit = twint_configs[0].Limit

    class MockRpcInitStub:
        def __init__(self, _mp_queue: queue.Queue):
            self._mp_queue = _mp_queue

        def StoreTweetsStream(self, iter_pb2_tweets_responses: Iterator[StoreTweetsResponse]):
            i_tweet = 0
            for i_tweet, pb2_tweet in enumerate(iter_pb2_tweets_responses, start=1):
                self._mp_queue.put(pb2_tweet)
            return StoreTweetsResponse(nb_tweets_received=i_tweet,
                                       nb_tweets_stored=i_tweet)

    mock_rpc_init_stub = mocker.patch('storage.processors.twint_with_multiprocess.rpc_init_stub')
    # Need to use a MultiProcess (synchronized) Queue
    mp_manager = multiprocessing.Manager()
    mp_queue = mp_manager.Queue()
    mock_rpc_init_stub.return_value = MockRpcInitStub(mp_queue)

    # https://docs.pytest.org/en/latest/logging.html
    with caplog.at_level(logging.DEBUG):
        run_twint_with_multiprocessing(
            twint_configs,
            # dummy connection parameters to instantiate
            rpc_storage_addr="Test",
            rpc_storage_port=12345,
            # a connection to rpc storage server
            nb_tweets_by_chunk=nb_tweets_by_chunk,
            log_level="debug"
        )

    nb_twitter_api_requests = math.ceil(twint_limit / float(nb_tweets_by_chunk))
    assert len(caplog.records) == nb_twitter_api_requests

    # Generator on tweets saved in MP Queue
    @dataclass
    class GenPb2Tweets:
        queue: queue.Queue

        def __iter__(self):
            return self

        def __next__(self):
            while True:
                try:
                    pb2_tweets = self.queue.get_nowait()
                    break
                except queue.Empty:
                    raise StopIteration
            return pb2_tweets

    store_tweets_request = list(GenPb2Tweets(mp_queue))  # type: List[StoreTweetsRequest]
    assert len(store_tweets_request) == nb_tweets_by_chunk * nb_twitter_api_requests * len(twint_configs)

    tweets_by_users = defaultdict(list)
    for store_tweet_request in store_tweets_request:
        tweet = store_tweet_request.tweet
        tweets_by_users[tweet.user_name].append(tweet)
    assert sorted(list(map(str.lower, tweets_by_users.keys()))) == sorted(map(str.lower, twitter_users))


src/storage/processors/twint_with_mono_thread.py
import logging

import twint
from google.protobuf.json_format import MessageToDict
from pyconfr_2019.grpc_nlp.protos import StorageService_pb2, StorageService_pb2_grpc
from pyconfr_2019.grpc_nlp.tools.rpc_init_stub import rpc_init_stub

from storage.processors.twint_with_multiprocess import _convert_tweet_from_twint_to_pb2

logger = logging.getLogger(__name__)


def run_mono_thread_twint(twint_config: twint.Config, storage_rpc_stub):
    """

    Args:
        twint_config:
        storage_rpc_stub:

    Returns:

    """
    tweets_list = []

    twint_config.Store_object = True
    twint_config.Store_object_tweets_list = tweets_list
    twint.run.Search(twint_config)

    if not tweets_list:
        logger.error(f"No tweets found from twitter_user={twint_config.Username}")
        return

    def _stream_tweets():
        for tweet in tweets_list:
            yield StorageService_pb2.StoreTweetsRequest(
                tweet=_convert_tweet_from_twint_to_pb2(tweet)
            )

    store_tweets_stream_response = MessageToDict(
        storage_rpc_stub.StoreTweetsStream(_stream_tweets()),
        including_default_value_fields=True
    )
    logger.debug(f"Response(StoreTweetsStream) = {store_tweets_stream_response}")


def run_twint_in_mono_thread(twint_configs, twitter_analyzer_storage_addr, twitter_analyzer_storage_port):
    # Init gRPC services Storage
    storage_rpc_stub = rpc_init_stub(twitter_analyzer_storage_addr,
                                     twitter_analyzer_storage_port,
                                     StorageService_pb2_grpc.StorageServiceStub,
                                     service_name='[twint-mono_thread] twitter analyzer storage')
    for twint_config in twint_configs:
        run_mono_thread_twint(twint_config, storage_rpc_stub)


src/storage/processors/twint_with_multiprocess.py
import builtins
import logging
import multiprocessing
import queue
from collections import Counter
from contextlib import ExitStack
from dataclasses import dataclass, field
from functools import partial
from multiprocessing import Pool as MPPool
from typing import Any, Callable, Optional

import twint
from google.protobuf.json_format import MessageToDict
from mock import patch
from pyconfr_2019.grpc_nlp.protos import StorageService_pb2, StorageService_pb2_grpc, Tweet_pb2
from pyconfr_2019.grpc_nlp.protos.StorageService_pb2_grpc import StorageServiceStub
from pyconfr_2019.grpc_nlp.tools.fct_logger import init_logger
from pyconfr_2019.grpc_nlp.tools.grouper import grouper_it
from pyconfr_2019.grpc_nlp.tools.rpc_init_stub import rpc_init_stub
from pyconfr_2019.grpc_nlp.tools.timestamps import tweet_datetime_to_utc_timestamp

logger = logging.getLogger(__name__)


def run_twint_with_multiprocessing(
        twint_configs,
        rpc_storage_addr,
        rpc_storage_port,
        nb_tweets_by_chunk=20,
        log_level: Optional[str] = None
):
    """

    Args:
        twint_configs:
        rpc_storage_addr:
        rpc_storage_port:
        nb_tweets_by_chunk:
        log_level:

    Returns:

    """
    mp_manager = multiprocessing.Manager()
    mp_queue = mp_manager.Queue()
    mp_event_twint_worker_is_finish = mp_manager.Event()

    # declare (only) one consumer for storing twint's tweets into db (though gRPC (storage) service).
    mp_worker_consumer = multiprocessing.Process(
        name="consumer: tweets from twint",
        target=loop_consume_tweets_from_twint_mp,
        kwargs={
            'queue_sync_with_twint': mp_queue,
            'event_twint_worker_is_finish': mp_event_twint_worker_is_finish,
            'twitter_analyzer_storage_addr': rpc_storage_addr,
            'twitter_analyzer_storage_port': rpc_storage_port,
            'nb_tweets_by_chunk': nb_tweets_by_chunk,
            'log_level': log_level,
        }
    )
    # This is critical! The consumer function has an infinite loop
    # Which means it will never exit unless we set daemon to true
    mp_worker_consumer.daemon = True
    mp_worker_consumer.start()

    # for each twint config (i.e twint tweets producer)
    # link to the same multiprocessing queue
    for twint_config in twint_configs:
        twint_config.Store_object = True
        twint_config.Store_object_tweets_list = mp_queue

    # Init a multiprocessing pool for all twint configs (producers)
    # Launch worker on search (tweets) method
    nb_producers = len(twint_configs)
    mp_pool_producers = MPPool(nb_producers)
    _ = mp_pool_producers.map(
        partial(worker_on_twint_run_search,
                use_capture_printouts=True,
                log_level=log_level),  # argument apply to all calls
        twint_configs  # map on all twint configurations
    )
    mp_pool_producers.close()
    # wait until all producers finish
    mp_pool_producers.join()
    logger.debug("All producers (nb=%d) are finished", nb_producers)

    # set the event (flag) to notify all consumers that twint producers are finished
    mp_event_twint_worker_is_finish.set()
    # wait until all consumers finished
    mp_worker_consumer.join()
    logger.debug("All consumers (nb=1) are finished")

    assert mp_queue.qsize() == 0


def loop_consume_tweets_from_twint_mp(
        queue_sync_with_twint: queue.Queue,
        twitter_analyzer_storage_addr,
        twitter_analyzer_storage_port,
        event_twint_worker_is_finish,
        nb_tweets_by_chunk,
        log_level: Optional[str] = None
):
    """
    AttributeError: Can't pickle local object '...'
    https://github.com/ouspg/trytls/issues/196#issuecomment-239676366

    https://github.com/grpc/grpc/tree/master/examples/python/multiprocessing
    https://github.com/grpc/grpc/blob/master/examples/python/multiprocessing/client.py

    Args:
        queue_sync_with_twint:
        twitter_analyzer_storage_addr:
        twitter_analyzer_storage_port:
        event_twint_worker_is_finish:
        nb_tweets_by_chunk:
        log_level:

    Returns:

    """
    if log_level:
        init_logger(log_level)

    # init gRPC stub to access to Storage service(r)
    storage_rpc_stub = None
    if twitter_analyzer_storage_addr and twitter_analyzer_storage_port:
        storage_rpc_stub = rpc_init_stub(
            twitter_analyzer_storage_addr,
            twitter_analyzer_storage_port,
            StorageService_pb2_grpc.StorageServiceStub,
            service_name='[twint-multiprocessing] twitter analyzer storage'
        )

    def _func_exit_loop():
        #
        # To create code that needs to wait for all queued tasks to be
        # completed, the preferred technique is to use the join() method.
        # return (event_twint_worker_is_finish.is_set()
        #         and queue_sync_with_twint.qsize() == 0)
        return event_twint_worker_is_finish.is_set()

    loop_store_tweets(queue_sync_with_twint,
                      storage_rpc_stub,
                      func_is_time_to_exit=_func_exit_loop,
                      func_apply_to_input=_convert_tweet_from_twint_to_pb2,
                      nb_tweets_by_chunk=nb_tweets_by_chunk)


def worker_on_twint_run_search(
        twint_config,
        use_capture_printouts: bool = False,
        log_level: Optional[str] = None,
):
    # https://stackoverflow.com/questions/26923050/how-do-i-name-the-processes-in-a-multiprocessing-pool
    multiprocessing.current_process().name = f"producer: twint's tweets from <{twint_config.Username}>"

    if log_level:
        init_logger(log_level)

    def mock_logger_debug(_, *__, **___):
        pass

    # https://stackoverflow.com/questions/3024925/create-a-with-block-on-several-context-managers
    # https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack
    ctx_managers_capture_outputs = [
        patch.object(logging, 'debug', mock_logger_debug),
        patch.object(builtins, 'print', mock_logger_debug)
    ]

    with ExitStack() as stack:
        for mgr in ctx_managers_capture_outputs if use_capture_printouts else []:
            stack.enter_context(mgr)
        # Patch the entry point "append" for twint's store object container.
        # The patch connect append to put, and allow to use multiprocessing.Queue
        setattr(twint_config.Store_object_tweets_list, "append",
                twint_config.Store_object_tweets_list.put)
        logger.debug("Starting worker on (twint) searching tweets from twitter user=<%s>", twint_config.Username)
        twint.run.Search(twint_config)
        logger.debug("Stopping worker on (twint) searching tweets from twitter user=<%s>", twint_config.Username)


def _convert_tweet_from_twint_to_pb2(tweet) -> Tweet_pb2.Tweet:
    return Tweet_pb2.Tweet(
        created_at=tweet_datetime_to_utc_timestamp(tweet.datetime, tweet.timezone),
        text=tweet.tweet,
        user_id=tweet.user_id,
        lang='',
        tweet_id=tweet.id,
        user_name=tweet.username,
    )


def loop_store_tweets(
        queue_with_pb2_tweets: queue.Queue,
        storage_rpc_stub: Optional[StorageServiceStub],
        func_is_time_to_exit: Callable[[], bool],
        func_apply_to_input: Callable[[Any], Any] = lambda i: i,
        nb_tweets_by_chunk: int = 50,
):
    def _stream_chunk_tweets():
        # https://anandology.com/python-practice-book/iterators.html
        @dataclass
        class GenPb2Tweets:
            queue: queue.Queue
            #
            timeout: float = field(default=1.00)

            def __iter__(self):
                return self

            def __next__(self):
                while True:
                    try:
                        pb2_tweets = func_apply_to_input(self.queue.get(timeout=self.timeout))
                        break
                    except queue.Empty:
                        if func_is_time_to_exit():
                            raise StopIteration
                return pb2_tweets

        for _chunk_tweets in grouper_it(GenPb2Tweets(queue_with_pb2_tweets),
                                        nb_tweets_by_chunk):
            yield _chunk_tweets

    for chunk_tweets in _stream_chunk_tweets():
        tweets = [
            StorageService_pb2.StoreTweetsRequest(tweet=tweet)
            for tweet in chunk_tweets
        ]

        def _format_debug_infos():
            return ", ".join(
                f"{user_name}#{counter}"
                for user_name, counter in Counter([tweet.tweet.user_name.lower() for tweet in tweets]).items()
            )

        logger.debug("gRPC: store %d tweets from users=(%s) in db ...", len(tweets), _format_debug_infos())

        if storage_rpc_stub:
            # -> gRPC Storage service
            store_tweets_stream_response = MessageToDict(
                storage_rpc_stub.StoreTweetsStream(iter(tweets)),
                including_default_value_fields=True
            )
            logger.debug(f"Response(StoreTweetsStream) = {store_tweets_stream_response}")
        # tweets from queue stored in database, so the queue task is done (for this chunk)
        queue_with_pb2_tweets.task_done()

    # logger.debug("nb_check_executed_for_exit=%d", nb_check_executed_for_exit)


src/storage/client_rpc_storage_from_twitter_into_db.py
#!/usr/bin/env python
"""
    -u EmmanuelMacron Trump PyCon OpenGL Lyon Paris Marseille Bordeaux Toulouse

    # multiprocessing with linux shell and no database storing/saving.
    ╰─ time (twint -u EmmanuelMacron --limit 500 & twint -u Trump --limit 500 & twint -u PyCon --limit 500 & twint -u OpenGL --limit 500 & wait)
    ( twint -u EmmanuelMacron --limit 500 & twint -u Trump --limit 500 & twint -u)  28.56s user 2.92s system 103% cpu 30.502 total

    --process run_twint_with_multiprocessing -u EmmanuelMacron Trump PyCon -l 500  --log_level info
    2019-09-01 12:33:19 - client_rpc_storage_from_twitter_into_db - INFO - Time elapsed: 32.2337887

    TODO:
      - write unit tests ! :p
"""
import argparse
import logging
import os
import signal
import sys
from timeit import default_timer as timer

import twint
from pyconfr_2019.grpc_nlp.tools.fct_logger import init_logger

from storage.processors.twint_with_cli import run_twint_with_cli
from storage.processors.twint_with_multiprocess import run_twint_with_multiprocessing

logger = logging.getLogger(__name__)

SIGNALS = [signal.SIGINT, signal.SIGTERM]


def process(args: argparse.Namespace):
    def _signal_handler(_sig, _):
        """ Empty signal handler used to override python default one """
        logger.info("sig: {} intercepted. Closing application.".format(_sig))
        # https://stackoverflow.com/questions/73663/terminating-a-python-script
        sys.exit()

    # Signals HANDLER (to exit properly)
    for sig in SIGNALS:
        signal.signal(sig, _signal_handler)

    twint_configs = []
    for twitter_user in args.twitter_users:
        twint_config = twint.Config()
        twint_config.Username = twitter_user
        twint_config.Limit = args.twint_limit  # bug with the Limit parameter not working only factor of 25 tweets
        twint_config.Debug = args.twint_debug
        twint_configs.append(twint_config)

    twitter_analyzer_storage_host = (args.twitter_analyzer_storage_addr,
                                     args.twitter_analyzer_storage_port)
    if args.processor == 'run_twint_with_cli':
        run_twint_with_cli(twint_configs)
    elif args.processor == 'run_twint_with_multiprocessing':
        run_twint_with_multiprocessing(twint_configs, *twitter_analyzer_storage_host, log_level=args.log_level)


def build_parser(parser=None, **argparse_options):
    """
    Args:
        parser (argparse.ArgumentParser):
        **argparse_options (dict):
    Returns:
    """
    if parser is None:
        parser = argparse.ArgumentParser(**argparse_options)

    argparse_default = "(default=%(default)s)."

    # TODO: grab parser from twint lib and getting all options available (by inheritance)
    parser.add_argument('-p', '--processor',
                        dest='processor',
                        type=str,
                        choices=['run_twint_with_multiprocessing',
                                 'run_twint_with_cli'],
                        default='run_twint_with_multiprocessing',
                        help=f"Processor to use to grab tweets (from Twinter). {argparse_default}")
    parser.add_argument('-u', '--twitter_users',
                        dest='twitter_users',
                        nargs='+',
                        type=str,
                        required=True,
                        help="Twitter user.")

    parser.add_argument("-l", "--twint_limit",
                        dest="twint_limit",
                        type=int,
                        required=False,
                        default=100,
                        help=f"Twint limit for scrapping tweets. {argparse_default}")

    parser.add_argument("--twint_debug",
                        dest="twint_debug",
                        action="store_true",
                        help="Activate Debug mode for twint (printout results in logs).")
    # GRPC SERVICES
    # - STORAGE
    parser.add_argument("--twitter_analyzer_storage_addr",
                        default=os.environ.get('GRPC_STORAGE_SERVICE_HOST', 'localhost'),
                        type=str,
                        help=f"{argparse_default}")
    parser.add_argument("--twitter_analyzer_storage_port",
                        default=int(os.environ.get('GRPC_STORAGE_PORT', '50052')),
                        type=int,
                        help=f"{argparse_default}",
                        metavar='')

    parser.add_argument(
        '-ll', '--log_level',
        type=str, required=False, default='debug',
        choices=['debug', 'warning', 'info', 'error', 'critical'],
        help=f"The logger filter level. {argparse_default}",
    )
    parser.add_argument(
        '-lf', '--log_file',
        type=str, required=False,
        help="The path to the file into which the logs will be streamed. "
             f"{argparse_default}",
    )

    parser.add_argument("-v", "--verbose",
                        action="store_true", default=False,
                        help="increase output verbosity (enable 'DEBUG' level log). "
                             f"{argparse_default}")

    return parser


def parse_arguments(args=None):
    """
    Returns:
        argparse.Namespace:
    """
    return build_parser().parse_args(args)


def main(args=None):
    start = timer()

    # Deal with inputs (stdin, parameters, etc ...)
    args = parse_arguments(args)

    init_logger(args.log_level)

    process(args)

    end = timer()
    logger.info(f"Time elapsed: {end - start}")


if __name__ == '__main__':
    main()
    sys.exit(0)
